rfcs/0015-consistent-environment-naming.md

```diff
@@ -0,0 +1,84 @@
+- Feature Name: consistent-environment-naming
+- Start Date: 2022-09-14
+- RFC PR: [mbta/technology-docs#0015](https://github.com/mbta/technology-docs/pull/15)
+- Asana task: [Dev environment naming / usage conventions](https://app.asana.com/0/1200506724882024/1202675232761300/f)
+- Status: Proposed
+
+# Summary
+[summary]: #summary
+
+Non-production environments of CTD applications have clear, descriptive, and consistent naming.
+
+# Motivation
+[motivation]: #motivation
+
+Non-production environments for our applications are generally named with some variation on `dev`. This doesn't capture the distinctions between different non-production uses of application environments.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Environments are named in `kebab-case`, following the convention `[name of application]-[modifier]`. Available modifiers are as follows, with a brief description of their use case:
+
+- `prod`: the production environment.
+- `staging`: used for final integration testing and monitoring for errors before deploying changes to production. Often, merging a PR into the `main` / `master` branch will be set up to automatically initiate a deploy to the application's `staging` environment. There should be one stage environment per application.
+- `test`: used as the target environment for some sort of recurring automated testing, such as end-to-end web application testing or load testing. Accepts an optional modifier specifying what testing tool the environment is for (for instance, `test-cypress` or `test-locust`). This is presently a hypothetical use case, albeit one that has been discussed from time to time.
+- `dev-[color]`: for experimenting with and evaluating features under active development, generally in their own branch. Potential uses include: allowing a developer to experiment with code that relates to functionality that's hard to replicate locally, providing a space for other team members to evaluate and sign off on a feature before it is merged, and collecting application-specific metrics on a particular change before deciding whether to merge it. `[color]` can be the color of any MBTA rapid transit line (`green`, `red`, `orange`, `blue`, or `silver`).
```

Comment:
question: if in the future we were to have fully ephemeral environments (i.e. the ability for a deployment to spin up a new environment itself for testing a feature) would this naming scheme still work? I worry that we might run out of transit lines and I don't want to artificially limit the number of app environments we can have. Additionally, we're currently discussing manually spinning up a new prediction_analyzer environment just for running a specific experiment against Swiftly data. This environment would only exist for the duration of the experiment, as a way of being able to maintain full `prediction_analyzer` capabilities for current service and the experiment concurrently. I would prefer to call this environment `swiftly-experiment` or something like that, would you advise giving it a line color like `dev-green` instead?

Reply:
Also very good questions. This is definitely written under the assumption that dev environments are something that you spin up and that then stick around for a long time, and you reuse them for a variety of purposes as time goes on. I think that describes our current paradigm of operating pretty well, but that doesn't necessarily need to be the case in the future. Maybe I can add a clause along the lines of "...unless it's strictly for testing a specific piece of functionality, in which case feel free to give it a name that indicates what the feature in question is."

Reply:
One of the benefits of Terraform was to make it easy to spin up single-use environments like this, and I expect it will happen more than once in the future. I think it should still have a prefix, so `dev-swiftly-experiment` or `tmp-swiftly-experiment` or something.

Reply:
I pushed a change reflecting that names specifying the feature being tested are allowed. I think they should still be `dev-` rather than `tmp-` or something else like that because `dev-` reflects the functional use of the environment. Separately I'd argue that our Terraform setup isn't quite there yet in terms of making it easy to set up a temporary environment for a feature that's under development for less than a few weeks or so, but that's definitely something we should accommodate in the future.